staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -319,12 +349,19 @@ func (w *watchCache) waitUntilFreshAndBlock(resourceVersion uint64, trace *utilt
 }
 
 // WaitUntilFreshAndList returns list of pointers to <storeElement> objects.
-func (w *watchCache) WaitUntilFreshAndList(resourceVersion uint64, trace *utiltrace.Trace) ([]interface{}, uint64, error) {
+func (w *watchCache) WaitUntilFreshAndList(resourceVersion uint64, matchValues []storage.MatchValue, trace *utiltrace.Trace) ([]interface{}, uint64, error) {
 	err := w.waitUntilFreshAndBlock(resourceVersion, trace)
 	defer w.RUnlock()
 	if err != nil {
 		return nil, 0, err
 	}
+
+	// TODO: use the smallest key size index.
+	for _, matchValue := range matchValues {
```

Comment:
This appears to find the first one with a match and return only that. Is that the right behavior? I.e. is it the case that objects must match all the values, so that it doesn't matter which index we search? Please add a comment explaining.

Reply:
Yes, the objects should match all the values. So the objects indexed by field and label will be filtered by field and label selectors again in later. In future, we will compare different MatchValues, return the  least objects

Reply:
Let me clarify one thing - this isn't the place where we do "final filtering" - only some "prefiltering" is happening here.
The final filtering is happening here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L720

So the only requirement here is to NOT miss anything that should be returned. We can return as many non-matching items as we want - they will be filtered out later. The fact that we return less things is only further performance improvement.

Reply:
@shaloulcy - it's probably worth adding a comment about that (something along the lines what I wrote above)

Reply:
done